servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java

```diff
@@ -100,28 +103,43 @@ public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt
                         }
                     });
 
-                    DeferSslHandler deferSslHandler = channel.pipeline().get(DeferSslHandler.class);
+                    final DeferSslHandler deferSslHandler = channel.pipeline().get(DeferSslHandler.class);
                     if (deferSslHandler == null) {
-                        return response.payloadBodyAndTrailers().ignoreElements().concat(failed(
-                                new IllegalStateException("Failed to find a handler of type " +
-                                        DeferSslHandler.class + " in channel pipeline.")));
+                        return failed(new IllegalStateException("Failed to find a handler of type " +
+                                DeferSslHandler.class + " in channel pipeline."));
                     }
-
                     deferSslHandler.ready();
 
-                    // There is no need to apply offloading explicitly (despite completing `processor` on the
-                    // EventLoop) because `payloadBody()` will be offloaded according to the strategy for the
-                    // request.
-                    return response.payloadBodyAndTrailers().ignoreElements().concat(fromSource(processor));
-                } else {
-                    return response.payloadBodyAndTrailers().ignoreElements().concat(
-                            failed(new ProxyResponseException("Bad response from proxy CONNECT " + connectAddress,
-                                    response.status())));
-                }
-            }));
+                    return fromSource(processor);
+                    // Close recently created connection in case of any error or cancellation while it connects to proxy
+                 }).whenFinally(new TerminalSignalConsumer<C>() {
+                    @Override
+                    public void onSuccess(@Nullable final C result) {
+                        // noop
+                    }
+
+                    @Override
+                    public void onError(final Throwable ignore) {
+                        closeConnection(c);
+                    }
+
+                    @Override
+                    public void cancel() {
+                        closeConnection(c);
+                    }
+                 });
+            } catch (Exception e) {
+                closeConnection(c);
+                return failed(e);
+            }
+            });
         }
     }
 
+    private static <C extends FilterableStreamingHttpConnection> void closeConnection(C connection) {
+        connection.closeAsync().subscribe();
```

Comment:
Same as above, we should not be doing dangling subscribes, instead compose connection closure with error response.
Cancel path will be an exception to this rule as cancels are not composable.

Reply:
Good catch! Two questions:
1. If we concat with failed Single, there is a chance that `connection.closeAsync()` may also fail. Should we worry about handling that case and adding the original exception as a suppressed one for the failed closure?
2. Should we make the same change for RRLB as a follow-up? Currently, we just subscribe to closeAsync and return an exception. For example: https://github.com/apple/servicetalk/blob/52532d402b8f7e33670942794f54388cae6844c0/servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java#L279-L285

Reply:
A1: As `closeAsync()` is idempotent and not expected to fail I will avoid handling that case and make the control flow complex here.

A2: Good catch, yes please do a follow-up

Reply:
Follow-up for RRLB: https://github.com/apple/servicetalk/pull/1006